File: 360/solution.go
package main

import (
	"fmt"
	"slices"
)

func main() {
	expect([]int{1, 0, 0, 0, 1}, 1, true)
	expect([]int{1, 0, 0, 0, 1}, 2, false)
	expect([]int{0, 0, 0, 0, 0}, 3, true)
	expect([]int{1, 0, 1, 0, 1}, 1, false)

	// Extra cases
	expect([]int{0, 0, 0}, 2, true)
	expect([]int{0}, 1, true)
	expect([]int{0}, 3, false)
}

func expect(field []int, k int, expected bool) {
	actual := canPlantFlowers(field, k)
	if actual != expected {
		fmt.Printf("Expected canPlantFlowers(%v) == %v, got %v\n", field, expected, actual)
	}
}

const SOIL = 0
const PLANT = 1

func canPlantFlowers(field []int, k int) bool {
	// Clone the slice
	field = slices.Clone(field)
	for i, plantOrSoil := range field {
		if plantOrSoil == PLANT {
			continue
		}
		if (i == 0 || field[i-1] == SOIL) && (i+1 == len(field) || field[i+1] == SOIL) {
			field[i] = PLANT
			// Update the number of flowers to plant
			k--
		}
	}

	return k == 0
}

File: 360/solution.js
main();

function main() {
  expect([1, 0, 0, 0, 1], 1, true);
  expect([1, 0, 0, 0, 1], 2, false);
  expect([0, 0, 0, 0, 0], 3, true);
  expect([1, 0, 1, 0, 1], 1, false);
  
  // Extra cases
  expect([0, 0, 0], 2, true);
  expect([0], 1, true);
  expect([0], 3, false);
}

function expect(field, k, expected) {
  const actual = canPlantFlowers(field, k);
  if (actual != expected) {
    console.error(`Expected canPlantFlowers([${field}]) == ${expected}, got ${actual}`);
  }
}

function canPlantFlowers(field, k) {
  const PLANT = 1;
  const SOIL = 0;

  // Clone the array
  field = field.slice();
  let i = 0;
  let cur;
  while (i < field.length && k > 0) {
    cur = field[i];
    if (cur == PLANT) {
      i++;
      continue;
    }
    if (cur == SOIL && 
      (i == 0 || field[i-1] == SOIL) && 
      (i+1 == field.length || field[i+1] == SOIL)) {
      field[i] = PLANT;
      k--;
    }
    i++;
  }

  return k == 0;
}